Source/DafnyLanguageServer/Language/DafnyProgramVerifier.cs

```diff
@@ -27,10 +27,17 @@ public class DafnyProgramVerifier : IProgramVerifier {
       this.logger = logger;
     }
 
-    public async Task<IReadOnlyList<IImplementationTask>> GetVerificationTasksAsync(ExecutionEngine engine,
+    public async Task<IReadOnlyList<IImplementationTask>> GetVerificationTasksAsync(ExecutionEngine boogieEngine,
       CompilationAfterResolution compilation,
       ModuleDefinition moduleDefinition,
       CancellationToken cancellationToken) {
+      var engine = boogieEngine;
+      if (compilation.Project.UsesProjectFile &&
+          (compilation.Program.Reporter.Options.Get(CommonOptionBag.WarnContradictoryAssumptions)
+           || compilation.Program.Reporter.Options.Get(CommonOptionBag.WarnRedundantAssumptions)
+          )) { // We need a separate Boogie engine
+        engine = new ExecutionEngine(compilation.Program.Reporter.Options, engine.Cache);
+      }
```

Comment:
This is a crucial place: Before, the engine was never considering the options in dfyconfig.toml. These lines ensure that a new engine is spun if we use a project file. This is a temporary solution only for the experimental feature of warning about assumptions.
.
I think we might want a better long-term solution to avoid creating an execution engine for every thread; Did you consider that, @keyboardDrummer  ?

Reply:
Instead of creating a new Boogie engine here, we should move the field `ProjectManagerDatabase.boogieEngine` to `ProjectManager`. ProjectManager is the longest lifetime in which the final options are available.

Reply:
I'm slightly confused by this suggestion. It seems that `ProjectManager` already has a `boogieEngine` field (and has since July), and the value of that field is exactly what's being passed in to `GetVerificationTasksAsync` here.

Reply:
Yes it has that field but it does not create the BoogieEngine, which is a problem because now the BoogieEngine does not use the options specific to that ProjectManager.

Reply:
The reason I only instantiated a single BoogieEngine is because each BoogieEngine creates a number of large stack threads, because our stack-dependent code overflows the stack when using regular threads, and thus consumes a lot of memory, and if you don't use project files, one BoogieEngine is created per file open in the IDE.

So.. what we also need is a way to pass a custom large stack thread pool to ExecutionEngine when creating it, so we can share those pools between the engines.

The following ExecutionEngine constructor from Boogie should have a `TaskScheduler largeStackThreadScheduler` argument:
```
    public ExecutionEngine(ExecutionEngineOptions options, VerificationResultCache cache) {
      Options = options;
      Cache = cache;
      checkerPool = new CheckerPool(options);
      verifyImplementationSemaphore = new SemaphoreSlim(Options.VcsCores);
      
      largeThreadScheduler = CustomStackSizePoolTaskScheduler.Create(16 * 1024 * 1024, Options.VcsCores);
      largeThreadTaskFactory = new(CancellationToken.None, TaskCreationOptions.None, TaskContinuationOptions.None, largeThreadScheduler);
    }
```

Reply:
We shouldn't create more Boogie `ExecutionEngine`s than we need, since it also has a `CheckerPool`, which is a pool of solvers it uses to verify things. This pool is specific to the DafnyOptions, since these options can configure the solver. So.. we should create 1 `ExecutionEngine` per `ProjectManager`

Reply:
Alright, good. I like that. And I can make the change to do that as part of this PR (though I think the Boogie refactoring should wait for a later one).

Reply:
I see several things we could change in Boogie:
1. Allow passing a CustomStackSizePoolTaskScheduler to ExecutionEngine
1. Make CustomStackSizePoolTaskScheduler lazily create new threads
1. Make CustomStackSizePoolTaskScheduler scale up and down

I'd feel safer about creating 1 ExecutionEngine per ProjectManager if we do (1) as well, which should be a trivial change, then we don't consume any additional memory as part of this PR.